### PR TITLE
Pessimistic lock

### PR DESCRIPTION
**Pull request by angelo-rendina-prima**

### CODE REVIEW

File: examples/optimistic_lock/Cargo.toml [hunk 1]
```diff
@@ -0,0 +1,23 @@
+[package]
+authors = ["Oliver Browne <oliver.browne@prima.it"]
```
Comment (cottinisimone) [inline examples/optimistic_lock/Cargo.toml:2]: lol you can write your name :)

File: examples/optimistic_lock/Cargo.toml [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+description = "Simple esrs side-effect example"
```
Comment (cottinisimone) [inline examples/optimistic_lock/Cargo.toml:3]: Description?

File: src/esrs/postgres/store.rs [hunk 3]
```diff
@@ -201,6 +213,20 @@ where
 {
     type Manager = Manager;
 
+    async fn lock(&self, aggregate_id: Uuid) -> Result<EventStoreLockGuard, <Self::Manager as Aggregate>::Error> {
+        let (key, _) = aggregate_id.as_u64_pair();
+        let connection = self.inner.pool.acquire().await.expect("TODO");
+        let lock_guard = PgStoreLockGuardAsyncSendBuilder {
+            lock: PgAdvisoryLock::with_key(PgAdvisoryLockKey::BigInt(key as i64)),
+            guard_builder: |lock: &PgAdvisoryLock| {
+                Box::pin(async move { lock.acquire(connection).await.expect("TODO") })
```
Comment (cottinisimone) [inline src/esrs/postgres/store.rs:222]: TODO :)

File: Dockerfile [hunk 4]
```diff
@@ -1,4 +1,4 @@
-FROM public.ecr.aws/prima/rust:1.62.1-1
+FROM public.ecr.aws/prima/rust:1.64.0
```
Comment (cottinisimone) [inline Dockerfile:1]: Maybe now there's the1.65?

File: examples/pessimistic_lock/src/aggregate.rs [hunk 5]
```diff
@@ -0,0 +1,68 @@
+use serde::{Deserialize, Serialize};
```
Comment (cottinisimone) [inline examples/pessimistic_lock/src/aggregate.rs:1]: I think you can include another example, so you can avoid rewriting this aggregate stuff

File: Dockerfile (hunk 4 above)
Comment (angelo-rendina-prima) [reply Dockerfile:1]: Nope still 1.64

File: examples/pessimistic_lock/src/aggregate.rs (hunk 5 above)
Comment (angelo-rendina-prima) [reply examples/pessimistic_lock/src/aggregate.rs:1]: Eh all other examples have their own aggregate, I just went with the slimmest I could think of (avoiding projectors and policies).

**Review by miterst — APPROVED** [review]
LGTM. We can give it a try!

File: Dockerfile (hunk 4 above)
Comment (angelo-rendina-prima) [reply Dockerfile]: Actually just bumped 🚀

File: examples/pessimistic_lock/src/aggregate.rs (hunk 5 above)
Comment (angelo-rendina-prima) [reply examples/pessimistic_lock/src/aggregate.rs]: Done, recycled the aggregate from another example.

**Comment by angelo-rendina-prima** [comment]
Bumped Rust image to 1.65, had to make a tiny change around deref to satisfy the new version of clippy.
Renamed the example to `locking_strategies`, and added two extra snippets about atomic **reads**.

**Review by cottinisimone — APPROVED** [review]

**Pull request merged**